
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>PReLIM &#8212; PReLIM 1.0 documentation</title>
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <h1>Source code for PReLIM</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">__future__</span> <span class="k">import</span> <span class="n">print_function</span>

<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Author: </span>
<span class="sd">Jack Duryea</span>
<span class="sd">Waterland Lab</span>
<span class="sd">Computational Epigenetics Section</span>
<span class="sd">Baylor College of Medicine</span>

<span class="sd">PReLIM: Preceise Read Level Imputation of Methylation</span>

<span class="sd">PReLIM imputes missing CpG methylation</span>
<span class="sd">states in CpG matrices.</span>

<span class="sd">&quot;&quot;&quot;</span>

<span class="c1"># standard imports</span>
<span class="kn">from</span> <span class="nn">scipy</span> <span class="k">import</span> <span class="n">stats</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">from</span> <span class="nn">tqdm</span> <span class="k">import</span> <span class="n">tqdm</span>
<span class="kn">import</span> <span class="nn">copy</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">from</span> <span class="nn">random</span> <span class="k">import</span> <span class="n">shuffle</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="k">import</span> <span class="n">defaultdict</span>
<span class="kn">import</span> <span class="nn">random</span>

<span class="c1"># sklearn imports</span>
<span class="kn">from</span> <span class="nn">sklearn.preprocessing</span> <span class="k">import</span> <span class="n">normalize</span>
<span class="kn">from</span> <span class="nn">sklearn.model_selection</span> <span class="k">import</span> <span class="n">train_test_split</span>
<span class="kn">from</span> <span class="nn">sklearn.metrics</span> <span class="k">import</span> <span class="n">roc_curve</span><span class="p">,</span> <span class="n">auc</span>
<span class="kn">from</span> <span class="nn">sklearn.ensemble</span> <span class="k">import</span> <span class="n">RandomForestClassifier</span>
<span class="kn">from</span> <span class="nn">sklearn.model_selection</span> <span class="k">import</span> <span class="n">GridSearchCV</span>

<span class="c1"># Pickle</span>
<span class="k">try</span><span class="p">:</span>
	<span class="kn">import</span> <span class="nn">cPickle</span> <span class="k">as</span> <span class="nn">p</span>
<span class="k">except</span> <span class="n">ModuleNotFoundError</span><span class="p">:</span>
	<span class="kn">import</span> <span class="nn">pickle</span> <span class="k">as</span> <span class="nn">p</span>


<span class="c1"># warnings suck, turn them off</span>
<span class="k">if</span> <span class="n">sys</span><span class="o">.</span><span class="n">version_info</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">3</span><span class="p">:</span>
	<span class="n">warnings</span><span class="o">.</span><span class="n">simplefilter</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">,</span> <span class="ne">DeprecationWarning</span><span class="p">)</span>
	<span class="k">with</span> <span class="n">warnings</span><span class="o">.</span><span class="n">catch_warnings</span><span class="p">():</span>
		<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">,</span> <span class="n">category</span><span class="o">=</span><span class="ne">DeprecationWarning</span><span class="p">)</span>
		<span class="kn">import</span> <span class="nn">md5</span><span class="o">,</span> <span class="nn">sha</span>


<span class="c1"># TODO: most of these fields are redundant in our application</span>
<div class="viewcode-block" id="CpGBin"><a class="viewcode-back" href="../index.html#PReLIM.CpGBin">[docs]</a><span class="k">class</span> <span class="nc">CpGBin</span><span class="p">():</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	A class that contains information about a CpG Bin. Does not need </span>
<span class="sd">	to be used directly, PReLIM will use this class internally.</span>
<span class="sd">	&quot;&quot;&quot;</span>
	<span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> 
			<span class="n">matrix</span><span class="p">,</span> 
			<span class="c1">#relative_positions</span>
			<span class="n">binStartInc</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> 
			<span class="n">binEndInc</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> 
			<span class="n">cpgPositions</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> 
			<span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
			<span class="n">encoding</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> 
			<span class="n">missingToken</span><span class="o">=</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> 
			<span class="n">chromosome</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> 
			<span class="n">binSize</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> 
			<span class="n">species</span><span class="o">=</span><span class="s2">&quot;MM10&quot;</span><span class="p">,</span> 
			<span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> 
			<span class="n">tag1</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> 
			<span class="n">tag2</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
		<span class="sd">&quot;&quot;&quot; </span>
<span class="sd">		Constructor for a bin</span>

<span class="sd">		Inputs:</span>

<span class="sd">		matrix: numpy array, the bin&#39;s CpG matrix.</span>
<span class="sd">		binStartInc: integer, the starting, inclusive, chromosomal index of the bin.</span>
<span class="sd">		binEndInc: integer, the ending, inclusive, chromosomal index of the bin.</span>
<span class="sd">		cpgPositions: array of integers, the chromosomal positions of the CpGs in the bin.</span>
<span class="sd">		sequence: string, nucleotide sequence (A,C,G,T)</span>
<span class="sd">		encoding: array, a reduced representation of the bin&#39;s CpG matrix</span>
<span class="sd">		missingToken: integer, the token that represents missing data in the matrix.</span>
<span class="sd">		chromosome: string, the chromosome this bin resides in.</span>
<span class="sd">		binSize: integer, the number of base pairs this bin covers</span>
<span class="sd">		species: string, the speices this bin belongs too.</span>
<span class="sd">		verbose: boolean, print warnings, set to &quot;false&quot; for no error checking and faster speed</span>

<span class="sd">		tag1: anything, for custom use.</span>
<span class="sd">		tag2: anything, for custom use.</span>
<span class="sd">		&quot;&quot;&quot;</span>


		<span class="bp">self</span><span class="o">.</span><span class="n">cpgDensity</span> <span class="o">=</span> <span class="n">matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">readDepth</span> <span class="o">=</span> <span class="n">matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
		
		

		<span class="bp">self</span><span class="o">.</span><span class="n">matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">matrix</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">binStartInc</span> <span class="o">=</span> <span class="n">binStartInc</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">binEndInc</span> <span class="o">=</span> <span class="n">binEndInc</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">cpgPositions</span> <span class="o">=</span> <span class="n">cpgPositions</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">missingToken</span> <span class="o">=</span> <span class="n">missingToken</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">chromosome</span> <span class="o">=</span> <span class="n">chromosome</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">binSize</span> <span class="o">=</span> <span class="n">binSize</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">species</span> <span class="o">=</span> <span class="n">species</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">tag1</span> <span class="o">=</span> <span class="n">tag1</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">tag2</span> <span class="o">=</span> <span class="n">tag2</span></div>




<div class="viewcode-block" id="PReLIM"><a class="viewcode-back" href="../index.html#PReLIM.PReLIM">[docs]</a><span class="k">class</span> <span class="nc">PReLIM</span><span class="p">():</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	Class for a PReLIM model. </span>
<span class="sd">	&quot;&quot;&quot;</span>
	<span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cpgDensity</span><span class="o">=</span><span class="mi">2</span><span class="p">):</span>
		<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">		Constructor for a PReLIM model.</span>
<span class="sd">        :param cpgDensity: the density of the bins that will be used </span>
<span class="sd">        &quot;&quot;&quot;</span>

		<span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">=</span> <span class="kc">None</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">cpgDensity</span> <span class="o">=</span> <span class="n">cpgDensity</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">METHYLATED</span> <span class="o">=</span> <span class="mi">1</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">UNMETHYLATED</span> <span class="o">=</span> <span class="mi">0</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">MISSING</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">methylated</span> <span class="o">=</span> <span class="mi">1</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">unmethylated</span> <span class="o">=</span> <span class="mi">0</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">unknown</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>





	<span class="c1"># Train a model</span>
<div class="viewcode-block" id="PReLIM.train"><a class="viewcode-back" href="../index.html#PReLIM.PReLIM.train">[docs]</a>	<span class="k">def</span> <span class="nf">train</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">bin_matrices</span><span class="p">,</span> <span class="n">model_file</span><span class="o">=</span><span class="s2">&quot;no&quot;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
		<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">		Train a model</span>
<span class="sd">		</span>
<span class="sd">        :param bin_matrices: list of cpg matrices</span>
<span class="sd">        :param model_file: The name of the file to save the model to. If None, then create a file name that includes a timestamp. If you don&#39;t want to save a file, set this to &quot;no&quot;</span>
<span class="sd">        :param verbose: prints more info if true</span>
<span class="sd">        &quot;&quot;&quot;</span>

		<span class="n">X</span><span class="p">,</span><span class="n">y</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_X_y</span><span class="p">(</span><span class="n">bin_matrices</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
		
		<span class="c1"># Train the neural network model</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">X</span><span class="p">,</span><span class="n">y</span><span class="p">,</span> <span class="n">model_file</span><span class="o">=</span><span class="n">model_file</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span></div>
		



<div class="viewcode-block" id="PReLIM.fit"><a class="viewcode-back" href="../index.html#PReLIM.PReLIM.fit">[docs]</a>	<span class="k">def</span> <span class="nf">fit</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
			<span class="n">X_train</span><span class="p">,</span>
			<span class="n">y_train</span><span class="p">,</span>
			<span class="n">n_estimators</span> <span class="o">=</span> <span class="p">[</span><span class="mi">10</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">500</span><span class="p">,</span> <span class="mi">1000</span><span class="p">],</span>
			<span class="n">cores</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span>
			<span class="n">max_depths</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">30</span><span class="p">],</span>
			<span class="n">model_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
			<span class="n">verbose</span><span class="o">=</span><span class="kc">False</span>
			<span class="p">):</span>

		<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">		Train a random forest model using grid search on a feature matrix (X) and class labels (y)</span>

<span class="sd">		Usage: </span>
<span class="sd">		model.fit(X_train, y_train)	</span>

<span class="sd">        :param X_train: numpy array, Contains feature vectors.</span>
<span class="sd">        :param y_train: numpy array, Contains labels for training data.</span>
<span class="sd">        :param n_estimators: list, the number of estimators to try during a grid search.</span>
<span class="sd">        :param max_depths: list, the maximum depths of trees to try during a grid search.</span>
<span class="sd">        :param cores: integer, the number of cores to use during training, helpful for grid search.</span>
<span class="sd">        :param model_file:  string,The name of the file to save the model to. </span>
<span class="sd">			If None, then create a file name that includes a timestamp.</span>
<span class="sd">			If you don&#39;t want to save a file, set this to &quot;no&quot;</span>

<span class="sd">        :return: The trained sklearn model</span>
<span class="sd">        &quot;&quot;&quot;</span>

		<span class="n">grid_param</span> <span class="o">=</span> <span class="p">{</span>  
		 <span class="s2">&quot;n_estimators&quot;</span><span class="p">:</span> <span class="n">n_estimators</span><span class="p">,</span>
		  <span class="s2">&quot;max_depth&quot;</span><span class="p">:</span> <span class="n">max_depths</span><span class="p">,</span>
		<span class="p">}</span>

		<span class="c1"># Note: let the grid search use a lot of cores, but only use 1 for each forest</span>
		<span class="c1"># since dispatching can take a lot of time</span>
		<span class="n">rf</span> <span class="o">=</span> <span class="n">RandomForestClassifier</span><span class="p">(</span><span class="n">n_jobs</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">=</span> <span class="n">GridSearchCV</span><span class="p">(</span><span class="n">rf</span><span class="p">,</span> <span class="n">grid_param</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">cv</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">X_train</span><span class="p">,</span> <span class="n">y_train</span><span class="p">)</span>


		<span class="c1"># save the model</span>
		<span class="k">if</span> <span class="n">model_file</span> <span class="o">==</span> <span class="s2">&quot;no&quot;</span><span class="p">:</span>
			<span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span>

		<span class="k">if</span> <span class="ow">not</span> <span class="n">model_file</span><span class="p">:</span>
			<span class="n">model_file</span> <span class="o">=</span> <span class="s2">&quot;PReLIM_model&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">())</span>

		<span class="n">p</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="n">model_file</span><span class="p">,</span><span class="s2">&quot;wb&quot;</span><span class="p">))</span>

		<span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span></div>




	<span class="c1"># Feature collection directly from bins</span>
<div class="viewcode-block" id="PReLIM.get_X_y"><a class="viewcode-back" href="../index.html#PReLIM.PReLIM.get_X_y">[docs]</a>	<span class="k">def</span> <span class="nf">get_X_y</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">bin_matrices</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
		<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :param bin_matrices: list of CpG matrices</span>
<span class="sd">        :param verbose: prints more info if true</span>
<span class="sd">        :return: feature matrix (X) and class labels (y)</span>
<span class="sd">        &quot;&quot;&quot;</span>

		<span class="n">bins</span> <span class="o">=</span> <span class="p">[]</span>

		<span class="c1"># convert to bin objects for ease of use</span>
		<span class="k">for</span> <span class="n">matrix</span> <span class="ow">in</span> <span class="n">bin_matrices</span><span class="p">:</span>
			<span class="n">mybin</span> <span class="o">=</span> <span class="n">CpGBin</span><span class="p">(</span> <span class="n">matrix</span><span class="o">=</span><span class="n">matrix</span> <span class="p">)</span>
			<span class="n">bins</span><span class="o">.</span><span class="n">append</span><span class="p">(</span> <span class="n">mybin</span> <span class="p">)</span>
		
		<span class="c1"># find bins with no missing data</span>
		<span class="n">complete_bins</span> <span class="o">=</span> <span class="n">_filter_missing_data</span><span class="p">(</span> <span class="n">bins</span> <span class="p">)</span>
		<span class="n">shuffle</span><span class="p">(</span> <span class="n">complete_bins</span> <span class="p">)</span>
		
		<span class="c1"># apply masks</span>
		<span class="n">masked_bins</span> <span class="o">=</span> <span class="n">_apply_masks</span><span class="p">(</span> <span class="n">complete_bins</span><span class="p">,</span> <span class="n">bins</span> <span class="p">)</span>

		<span class="c1"># extract features</span>
		<span class="n">X</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_collectFeatures</span><span class="p">(</span> <span class="n">masked_bins</span> <span class="p">)</span> 
		<span class="k">return</span> <span class="n">X</span><span class="p">,</span> <span class="n">y</span> </div>


	<span class="c1"># Return a vector of predicted classes </span>
<div class="viewcode-block" id="PReLIM.predict_classes"><a class="viewcode-back" href="../index.html#PReLIM.PReLIM.predict_classes">[docs]</a>	<span class="k">def</span> <span class="nf">predict_classes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">X</span><span class="p">):</span>
		<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">		Predict the classes of the samples in the given feature matrix</span>
<span class="sd">		</span>
<span class="sd">		Usage:</span>
<span class="sd">		y_pred = CpGNet.predict_classes(X)  </span>
<span class="sd">        </span>
<span class="sd">        :param X: numpy array, contains feature vectors</span>
<span class="sd">        :param verbose: prints more info if true</span>
<span class="sd">        :return: 1-d numpy array of predicted classes</span>
<span class="sd">        &quot;&quot;&quot;</span>

		<span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">X</span><span class="p">)</span></div>
	
	<span class="c1"># Return a vector of probabilities for methylation</span>
<div class="viewcode-block" id="PReLIM.predict"><a class="viewcode-back" href="../index.html#PReLIM.PReLIM.predict">[docs]</a>	<span class="k">def</span> <span class="nf">predict</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">X</span><span class="p">):</span>
		<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">		Predict the probability of methylation for each sample in the given feature matrix</span>
<span class="sd">		</span>
<span class="sd">		Usage:</span>
<span class="sd">		y_pred = CpGNet.predict(X)  </span>
<span class="sd">        </span>
<span class="sd">        :param X: numpy array, contains feature vectors</span>
<span class="sd">        :param verbose: prints more info if true</span>
<span class="sd">        :return: 1-d numpy array of prediction values</span>
<span class="sd">        &quot;&quot;&quot;</span>

		<span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">predict_proba</span><span class="p">(</span><span class="n">X</span><span class="p">)[:,</span><span class="mi">1</span><span class="p">]</span></div>


<div class="viewcode-block" id="PReLIM.predict_proba"><a class="viewcode-back" href="../index.html#PReLIM.PReLIM.predict_proba">[docs]</a>	<span class="k">def</span> <span class="nf">predict_proba</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">X</span><span class="p">):</span>
		<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">		Predict the classes of the samples in the given feature matrix</span>
<span class="sd">		Same as predict, just a convenience to have in case of differen styles</span>
<span class="sd">		</span>
<span class="sd">		Usage:</span>
<span class="sd">		y_pred = CpGNet.predict_classes(X)  </span>
<span class="sd">        </span>
<span class="sd">        :param X: numpy array, contains feature vectors</span>
<span class="sd">        :param verbose: prints more info if true</span>
<span class="sd">        :return: 1-d numpy array of predicted classes</span>
<span class="sd">        &quot;&quot;&quot;</span>

		<span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">predict_proba</span><span class="p">(</span><span class="n">X</span><span class="p">)[:</span><span class="mi">1</span><span class="p">]</span></div>


	<span class="c1"># Load a saved model</span>
<div class="viewcode-block" id="PReLIM.loadWeights"><a class="viewcode-back" href="../index.html#PReLIM.PReLIM.loadWeights">[docs]</a>	<span class="k">def</span> <span class="nf">loadWeights</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">model_file</span><span class="p">):</span>
		<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">		self.model is loaded with the provided weights</span>

<span class="sd">		:param model_file: string, name of file with a saved model</span>
<span class="sd">		&quot;&quot;&quot;</span>

		<span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="n">model_file</span><span class="p">,</span><span class="s2">&quot;rb&quot;</span><span class="p">))</span></div>

	


	

	<span class="c1"># Imputes missing values in Bins</span>
<div class="viewcode-block" id="PReLIM.impute"><a class="viewcode-back" href="../index.html#PReLIM.PReLIM.impute">[docs]</a>	<span class="k">def</span> <span class="nf">impute</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">matrix</span><span class="p">):</span>
		<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">		Impute the missing values in a CpG matrix. Values are filled with the </span>
<span class="sd">		predicted probability of methylation.</span>

<span class="sd">		:param matrix: a 2d np array, dtype=float, representing a CpG matrix, 1=methylated, 0=unmethylated, -1=unknown</span>
<span class="sd">        :return: A 2d numpy array with predicted probabilities of methylation</span>
<span class="sd">		&quot;&quot;&quot;</span>

		<span class="n">X</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_imputation_features</span><span class="p">(</span><span class="n">matrix</span><span class="p">)</span>
		
		<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">X</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span> <span class="c1"># nothing to impute</span>
			<span class="k">return</span> <span class="n">matrix</span>

		<span class="n">predictions</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>

		<span class="n">k</span> <span class="o">=</span> <span class="mi">0</span> <span class="c1"># keep track of prediction index for missing states</span>
		<span class="n">predicted_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">matrix</span><span class="p">)</span>
		<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">predicted_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
			<span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">predicted_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
				<span class="k">if</span> <span class="n">predicted_matrix</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
					<span class="n">predicted_matrix</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">predictions</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>
					<span class="n">k</span> <span class="o">+=</span> <span class="mi">1</span>

		<span class="k">return</span> <span class="n">predicted_matrix</span></div>






	<span class="c1"># Extract all features for all matrices so we can predict in bulk, this is where the speedup comes from</span>
<div class="viewcode-block" id="PReLIM.impute_many"><a class="viewcode-back" href="../index.html#PReLIM.PReLIM.impute_many">[docs]</a>	<span class="k">def</span> <span class="nf">impute_many</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">matrices</span><span class="p">):</span>
		<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">		Imputes a bunch of matrices at the same time to help speed up imputation time.</span>
<span class="sd">	</span>
<span class="sd">		:param matrices: array-like (i.e. list), where each element is a 2d np array, dtype=float, representing a CpG matrix, 1=methylated, 0=unmethylated, -1=unknown</span>
<span class="sd">        :return: A List of 2d numpy arrays with predicted probabilities of methylation for unknown values.</span>
<span class="sd">		&#39;&#39;&#39;</span>
		
		<span class="n">X</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">features</span> <span class="k">for</span> <span class="n">matrix_features</span> <span class="ow">in</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_get_imputation_features</span><span class="p">(</span><span class="n">matrix</span><span class="p">)</span> <span class="k">for</span> <span class="n">matrix</span> <span class="ow">in</span> <span class="n">matrices</span><span class="p">]</span> <span class="k">for</span> <span class="n">features</span> <span class="ow">in</span> <span class="n">matrix_features</span><span class="p">])</span>
		
		<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">X</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
			<span class="k">return</span> <span class="n">matrices</span>
		
		<span class="n">predictions</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>


		<span class="n">predicted_matrices</span> <span class="o">=</span> <span class="p">[]</span>


		<span class="n">k</span> <span class="o">=</span> <span class="mi">0</span> <span class="c1"># keep track of prediction index for missing states, order is crucial!</span>
		<span class="k">for</span> <span class="n">matrix</span> <span class="ow">in</span> <span class="n">matrices</span><span class="p">:</span>
			<span class="n">predicted_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">matrix</span><span class="p">)</span>
			<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">predicted_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
				<span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">predicted_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
					<span class="k">if</span> <span class="n">predicted_matrix</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
						<span class="n">predicted_matrix</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">predictions</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>
						<span class="n">k</span> <span class="o">+=</span> <span class="mi">1</span>
			<span class="n">predicted_matrices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">predicted_matrix</span><span class="p">)</span>

		<span class="k">return</span> <span class="n">predicted_matrices</span></div>




	<span class="c1">### Helper functions, for private use only ###</span>

	<span class="c1"># get a feature matrix for the given cpg matrix</span>
	<span class="k">def</span> <span class="nf">_get_imputation_features</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">matrix</span><span class="p">):</span>
		<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">		Returns a vector of features needed for the imputation of this matrix</span>
<span class="sd">		Each sample is an individual CpG, and the features are</span>
<span class="sd">		the row mean, the column mean, the position of the cpg in the matrix,</span>
<span class="sd">		the row, and the relative proportions of each methylation pattern </span>

<span class="sd">		:param matrix: a 2d np array, dtype=float, representing a CpG matrix, 1=methylated, 0=unmethylated, -1=unknown</span>
<span class="sd">        :return: A feature vector for the matrix</span>
<span class="sd">		&#39;&#39;&#39;</span>
		
		<span class="n">X</span> <span class="o">=</span> <span class="p">[]</span>

		<span class="n">numReads</span> <span class="o">=</span> <span class="n">matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
		<span class="n">density</span> <span class="o">=</span> <span class="n">matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

		<span class="n">nan_copy</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">matrix</span><span class="p">)</span>
		<span class="n">nan_copy</span><span class="p">[</span><span class="n">nan_copy</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>

		<span class="c1"># get the column and row means</span>
		<span class="n">column_means</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="n">nan_copy</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
		<span class="n">row_means</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="n">nan_copy</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
		
		<span class="n">encoding</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_encode_input_matrix</span><span class="p">(</span><span class="n">matrix</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

		<span class="c1"># iterate over all values in the matrix</span>
		<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">numReads</span><span class="p">):</span>
			<span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">density</span><span class="p">):</span>
				<span class="n">observed_state</span> <span class="o">=</span> <span class="n">matrix</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span>
				
				<span class="c1"># only record missing values </span>
				<span class="k">if</span> <span class="n">observed_state</span> <span class="o">!=</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
					<span class="k">continue</span>

				<span class="n">row_mean</span> <span class="o">=</span> <span class="n">row_means</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
				<span class="n">col_mean</span> <span class="o">=</span> <span class="n">column_means</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
				
				<span class="n">row</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">matrix</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
				<span class="n">row</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>

				<span class="c1"># features for a single sample</span>
				<span class="n">data</span> <span class="o">=</span> <span class="p">[</span><span class="n">row_mean</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">col_mean</span><span class="p">]</span> <span class="o">+</span>  <span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">row</span><span class="p">)</span> <span class="o">+</span>  <span class="nb">list</span><span class="p">(</span><span class="n">encoding</span><span class="p">)</span>
				
				<span class="n">X</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

		<span class="c1"># list to np array</span>
		<span class="n">X</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>

		<span class="k">return</span> <span class="n">X</span>


	<span class="c1"># Returns a matrix encoding of a CpG matrix</span>
	<span class="k">def</span> <span class="nf">_encode_input_matrix</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">m</span><span class="p">):</span>
		<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">		:param m: a 2d np array, dtype=float, representing a CpG matrix, 1=methylated, 0=unmethylated, -1=unknown</span>
<span class="sd">        :return: list of relative proportions of each type of methylation pattern, number of reads</span>
<span class="sd">		&quot;&quot;&quot;</span>
		<span class="n">matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">m</span><span class="p">)</span>
		<span class="n">n_cpgs</span> <span class="o">=</span> <span class="n">matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
		<span class="n">matrix</span> <span class="o">+=</span> <span class="mi">1</span>  <span class="c1"># deal with -1s</span>
		<span class="n">base_3_vec</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">power</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">n_cpgs</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">))</span>
		
		<span class="n">encodings</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">base_3_vec</span><span class="p">,</span> <span class="n">matrix</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>
		
		<span class="n">encoded_vector_dim</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">power</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="n">n_cpgs</span><span class="p">)</span>
		<span class="n">encoded_vector</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">encoded_vector_dim</span><span class="p">)</span>
		
		<span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">encodings</span><span class="p">:</span>
			<span class="n">encoded_vector</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">)]</span> <span class="o">+=</span> <span class="mi">1</span>
		
		<span class="n">num_reads</span> <span class="o">=</span> <span class="n">encodings</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
		
		<span class="c1"># Now we normalize</span>
		<span class="n">encoded_vector_norm</span> <span class="o">=</span> <span class="n">normalize</span><span class="p">([</span><span class="n">encoded_vector</span><span class="p">],</span> <span class="n">norm</span><span class="o">=</span><span class="s2">&quot;l1&quot;</span><span class="p">)</span>
		<span class="k">return</span> <span class="n">encoded_vector_norm</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">num_reads</span>

	<span class="c1"># finds the majority class of the given column, discounting the current cpg</span>
	<span class="k">def</span> <span class="nf">_get_column_mean</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">matrix</span><span class="p">,</span> <span class="n">col_i</span><span class="p">,</span> <span class="n">current_cpg_state</span><span class="p">):</span>
		<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">		:param matrix: a 2d np array, dtype=float, representing a CpG matrix, 1=methylated, 0=unmethylated, -1=unknown</span>
<span class="sd">		:param col_i: integer, the column index</span>
<span class="sd">		:param current_cpg_state: the cpg to discount</span>
<span class="sd">        :return: the mean value of column col_i, discounting current_cpg_state</span>
<span class="sd">		&quot;&quot;&quot;</span>
		<span class="n">sub</span> <span class="o">=</span> <span class="n">matrix</span><span class="p">[:,</span> <span class="n">col_i</span><span class="p">]</span>
		<span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_mean</span><span class="p">(</span><span class="n">sub</span><span class="p">,</span> <span class="n">current_cpg_state</span><span class="p">)</span>

	<span class="c1"># finds the majority class of the given read, discounting the current cpg</span>
	<span class="k">def</span> <span class="nf">_get_read_mean</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">matrix</span><span class="p">,</span> <span class="n">read_i</span><span class="p">,</span> <span class="n">current_cpg_state</span><span class="p">):</span>
		<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">		:param matrix: a 2d np array, dtype=float, representing a CpG matrix, 1=methylated, 0=unmethylated, -1=unknown</span>
<span class="sd">		:param read_i: integer, the row index</span>
<span class="sd">		:param current_cpg_state: the cpg to discount</span>
<span class="sd">        :return: the mean value of row read_i, discounting current_cpg_state</span>
<span class="sd">		&quot;&quot;&quot;</span>
		<span class="n">sub</span> <span class="o">=</span> <span class="n">matrix</span><span class="p">[</span><span class="n">read_i</span><span class="p">,</span> <span class="p">:]</span>
		<span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_mean</span><span class="p">(</span><span class="n">sub</span><span class="p">,</span> <span class="n">current_cpg_state</span><span class="p">)</span>


	<span class="c1"># Return the mean of sub matrix, discounting the current cpg methylation state</span>
	<span class="k">def</span> <span class="nf">_get_mean</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sub_matrix</span><span class="p">,</span> <span class="n">current_cpg_state</span><span class="p">):</span>
		<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">		:param sub_matrix: a list of individual cpgs</span>
<span class="sd">		:param current_cpg_state: the cpg to discount</span>
<span class="sd">		:return: the mean value of the list, discounting current_cpg_state</span>
<span class="sd">		&#39;&#39;&#39;</span>
		<span class="n">num_methy</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">sub_matrix</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">METHYLATED</span><span class="p">)</span>
		<span class="n">num_unmethy</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">sub_matrix</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">UNMETHYLATED</span><span class="p">)</span>

		<span class="k">if</span> <span class="n">current_cpg_state</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">METHYLATED</span><span class="p">:</span>
			<span class="n">num_methy</span> <span class="o">-=</span> <span class="mi">1</span>
		<span class="n">num_methy</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">num_methy</span><span class="p">)</span>
		<span class="k">if</span> <span class="n">current_cpg_state</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">UNMETHYLATED</span><span class="p">:</span>
			<span class="n">num_unmethy</span> <span class="o">-=</span> <span class="mi">1</span>
		<span class="n">num_unmethy</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">num_unmethy</span><span class="p">)</span>
		<span class="k">if</span> <span class="nb">float</span><span class="p">(</span><span class="n">num_methy</span> <span class="o">+</span> <span class="n">num_unmethy</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
			<span class="k">return</span> <span class="o">-</span><span class="mi">2</span>

		<span class="k">return</span> <span class="nb">float</span><span class="p">(</span><span class="n">num_methy</span><span class="p">)</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="n">num_methy</span> <span class="o">+</span> <span class="n">num_unmethy</span><span class="p">)</span>

	<span class="c1"># Returns X, y</span>
	<span class="c1"># note: y can contain the labels 1,0, -1</span>
	<span class="k">def</span> <span class="nf">_collectFeatures</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">bins</span><span class="p">):</span>
		<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">		Given a list of cpg bins, collect features for each artificially masked CpG</span>
<span class="sd">		and record the hidden value as the class label.</span>

<span class="sd">		:param matrix: bins: list of CpG bins that contain CpG matrices</span>
<span class="sd">        :return: feature matrix X and class labels y</span>
<span class="sd">		&quot;&quot;&quot;</span>
		<span class="n">X</span> <span class="o">=</span> <span class="p">[]</span>
		<span class="n">Y</span> <span class="o">=</span> <span class="p">[]</span>
		<span class="k">for</span> <span class="n">Bin</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">bins</span><span class="p">):</span>
			<span class="n">observed_matrix</span> <span class="o">=</span> <span class="n">Bin</span><span class="o">.</span><span class="n">tag2</span><span class="p">[</span><span class="s2">&quot;observed&quot;</span><span class="p">]</span>
			<span class="n">truth_matrix</span> <span class="o">=</span> <span class="n">Bin</span><span class="o">.</span><span class="n">tag2</span><span class="p">[</span><span class="s2">&quot;truth&quot;</span><span class="p">]</span>
			<span class="n">encoding</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_encode_input_matrix</span><span class="p">(</span><span class="n">observed_matrix</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

			<span class="n">numReads</span> <span class="o">=</span> <span class="n">observed_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
			<span class="n">density</span> <span class="o">=</span> <span class="n">observed_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
			<span class="c1">#positions = Bin.cpgPositions</span>
			<span class="n">nan_copy</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">observed_matrix</span><span class="p">)</span>
			

			<span class="n">nan_copy</span><span class="p">[</span><span class="n">nan_copy</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span> 
			<span class="n">column_means</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="n">nan_copy</span><span class="p">,</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
			<span class="n">row_means</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="n">nan_copy</span><span class="p">,</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

			<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">numReads</span><span class="p">):</span>
				<span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">density</span><span class="p">):</span>
					<span class="n">observed_state</span> <span class="o">=</span> <span class="n">observed_matrix</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="n">j</span><span class="p">]</span>
					<span class="k">if</span> <span class="n">observed_state</span> <span class="o">!=</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
						<span class="k">continue</span>

					<span class="n">state</span> <span class="o">=</span> <span class="n">truth_matrix</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="n">j</span><span class="p">]</span>
					<span class="n">Y</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">state</span><span class="p">)</span>
					
					
					<span class="c1"># row and column means</span>
					<span class="n">row_mean</span> <span class="o">=</span> <span class="n">row_means</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
					<span class="n">col_mean</span> <span class="o">=</span> <span class="n">column_means</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>

					<span class="c1"># j is the current index in the row</span>
					
					<span class="c1"># encoding is the matrix encoding vector</span>
					<span class="c1"># differences is the difference in positions of the cpgs</span>
					<span class="n">row</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">observed_matrix</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
					<span class="n">row</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>

					<span class="n">data</span> <span class="o">=</span> <span class="p">[</span><span class="n">row_mean</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">col_mean</span><span class="p">]</span> <span class="o">+</span>  <span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">row</span><span class="p">)</span> <span class="o">+</span>  <span class="nb">list</span><span class="p">(</span><span class="n">encoding</span><span class="p">)</span>
					
					<span class="n">X</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>


		<span class="n">X</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>
		<span class="n">Y</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">Y</span><span class="p">)</span>
		<span class="n">Y</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
		<span class="k">return</span> <span class="n">X</span><span class="p">,</span> <span class="n">Y</span></div>



<span class="c1">#### Helper functions ####</span>

<span class="c1"># Returns a list of bins similar to the input</span>
<span class="c1"># but matrix rows with missing values are removed</span>
<span class="k">def</span> <span class="nf">_filter_bad_reads</span><span class="p">(</span><span class="n">bins</span><span class="p">):</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	Given a list of cpg bins, remove reads with missing values</span>
<span class="sd">	so we can mask them.</span>
<span class="sd">	</span>
<span class="sd">	:param matrix: bins: list of CpG bins that contain CpG matrices</span>
<span class="sd">    :return: bins, but all reads wiht missing values have been removed</span>
<span class="sd">	&quot;&quot;&quot;</span>
	<span class="n">filtered_bins</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="k">for</span> <span class="n">Bin</span> <span class="ow">in</span> <span class="n">bins</span><span class="p">:</span>
		<span class="n">newBin</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">Bin</span><span class="p">)</span>
		<span class="n">matrix</span> <span class="o">=</span> <span class="n">newBin</span><span class="o">.</span><span class="n">matrix</span>

		<span class="c1"># find rows with missing values</span>
		<span class="n">counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">matrix</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
		<span class="n">idx</span> <span class="o">=</span> <span class="n">counts</span> <span class="o">==</span> <span class="mi">0</span>
		<span class="n">matrix_filtered</span> <span class="o">=</span> <span class="n">matrix</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
		<span class="n">newBin</span><span class="o">.</span><span class="n">matrix</span> <span class="o">=</span> <span class="n">matrix_filtered</span>
		<span class="n">filtered_bins</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">newBin</span><span class="p">)</span>

	<span class="k">return</span> <span class="n">filtered_bins</span>

<span class="c1"># Returns a mapping of dimensions to list of masks that can be used on data</span>
<span class="c1"># of that size. the missing pattern is in matrix form.</span>
<span class="c1"># -1 is missing, 2 is known</span>
<span class="k">def</span> <span class="nf">_extract_masks</span><span class="p">(</span> <span class="n">bins</span><span class="p">):</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	Given a list of cpg bins, return a list matrices that</span>
<span class="sd">	represent the patterns of missing values, or &quot;masks&quot;</span>
<span class="sd">	</span>
<span class="sd">	:param matrix: bins: list of CpG bins that contain CpG matrices</span>
<span class="sd">    :return: list of matrices that represent the patterns of missing values</span>
<span class="sd">	&quot;&quot;&quot;</span>
	<span class="n">masks</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="k">lambda</span><span class="p">:</span> <span class="p">[])</span>
	<span class="k">for</span> <span class="n">Bin</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">bins</span><span class="p">):</span>
		<span class="n">matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">Bin</span><span class="o">.</span><span class="n">matrix</span><span class="p">)</span>
		<span class="n">matrix</span><span class="p">[</span><span class="n">matrix</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="mi">2</span>

		<span class="c1">#min_missing = 10</span>
		<span class="n">min_missing</span> <span class="o">=</span> <span class="mi">1</span> <span class="c1"># must have at least 1 missing value</span>
		<span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">matrix</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">min_missing</span><span class="p">:</span>
			<span class="n">masks</span><span class="p">[</span><span class="n">matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">matrix</span><span class="p">)</span>

	<span class="k">return</span> <span class="n">masks</span>

<span class="c1"># Extract masks from original matrices and apply them to the complete matrices</span>
<span class="k">def</span> <span class="nf">_apply_masks</span><span class="p">(</span> <span class="n">filtered_bins</span><span class="p">,</span> <span class="n">all_bins</span> <span class="p">):</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	Given a list of filtered cpg bins and a list of all the bins,</span>
<span class="sd">	extract masks from the original bins and apply them to the filtered bins.</span>
<span class="sd">	</span>
<span class="sd">	:param filtered_bins: bins with no reads with missing values.</span>
<span class="sd">	:param all_bins: list of CpG bins that contain CpG matrices</span>
<span class="sd">	:return: list of matrices that represent the patterns of missing values</span>
<span class="sd">	&quot;&quot;&quot;</span>
	<span class="n">masks</span> <span class="o">=</span> <span class="n">_extract_masks</span><span class="p">(</span> <span class="n">all_bins</span> <span class="p">)</span>
	<span class="n">ready_bins</span> <span class="o">=</span> <span class="p">[]</span>

	<span class="k">for</span> <span class="n">Bin</span> <span class="ow">in</span> <span class="n">filtered_bins</span><span class="p">:</span>
		<span class="n">truth_matrix</span> <span class="o">=</span> <span class="n">Bin</span><span class="o">.</span><span class="n">matrix</span>
		<span class="n">m_shape</span> <span class="o">=</span> <span class="n">truth_matrix</span><span class="o">.</span><span class="n">shape</span>
		<span class="k">if</span> <span class="n">m_shape</span> <span class="ow">in</span> <span class="n">masks</span><span class="p">:</span>
			<span class="k">if</span> <span class="nb">len</span><span class="p">(</span> <span class="n">masks</span> <span class="p">[</span> <span class="n">m_shape</span> <span class="p">]</span> <span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
				<span class="n">mask</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">masks</span><span class="p">[</span><span class="n">m_shape</span><span class="p">])</span>
				<span class="n">observed</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">minimum</span><span class="p">(</span><span class="n">truth_matrix</span><span class="p">,</span> <span class="n">mask</span><span class="p">)</span>
				<span class="n">Bin</span><span class="o">.</span><span class="n">tag2</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;truth&quot;</span><span class="p">:</span><span class="n">truth_matrix</span><span class="p">,</span> <span class="s2">&quot;observed&quot;</span><span class="p">:</span><span class="n">observed</span><span class="p">,</span> <span class="s2">&quot;mask&quot;</span><span class="p">:</span><span class="n">mask</span><span class="p">}</span>
				<span class="n">ready_bins</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Bin</span><span class="p">)</span>

	<span class="k">return</span> <span class="n">ready_bins</span>

<span class="c1"># Get a list of bins with no missing data</span>
<span class="k">def</span> <span class="nf">_filter_missing_data</span><span class="p">(</span> <span class="n">bins</span><span class="p">,</span> <span class="n">min_read_depth</span><span class="o">=</span><span class="mi">1</span> <span class="p">):</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	Given a list of filtered cpg bins and a list of all the bins,</span>
<span class="sd">	extract masks from the original bins and apply them to the filtered bins.</span>
<span class="sd">	</span>
<span class="sd">	:param bins: list of CpG bins that contain CpG matrices</span>
<span class="sd">	:param min_read_depth: minimum number of reads needed for a bin to be complete.</span>
<span class="sd">	:return: remove reads with missing values from bins</span>

<span class="sd">	&quot;&quot;&quot;</span>
	<span class="n">cpg_bins_complete</span> <span class="o">=</span> <span class="n">_filter_bad_reads</span><span class="p">(</span><span class="n">bins</span><span class="p">)</span>
	<span class="c1"># secondary depth filter</span>
	<span class="n">cpg_bins_complete_depth</span> <span class="o">=</span> <span class="p">[</span><span class="n">bin_</span> <span class="k">for</span> <span class="n">bin_</span> <span class="ow">in</span> <span class="n">cpg_bins_complete</span> <span class="k">if</span> <span class="n">bin_</span><span class="o">.</span><span class="n">matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">min_read_depth</span><span class="p">]</span>
	<span class="k">return</span> <span class="n">cpg_bins_complete_depth</span>


</pre></div>

          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">PReLIM</a></h1>








<h3>Navigation</h3>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="index.html">Module code</a><ul>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019, Jack Duryea.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
    </div>

    

    
  </body>
</html>